game/transfers.py

```diff
@@ -166,21 +166,49 @@ def find_escape_route(self) -> Optional[ControlPoint]:
             return self.transport.find_escape_route()
         return None
 
-    def proceed(self) -> None:
-        if not self.destination.is_friendly(self.player):
-            logging.info(f"Transfer destination {self.destination} was captured.")
-            if self.position.is_friendly(self.player):
-                self.disband_at(self.position)
-            elif (escape_route := self.find_escape_route()) is not None:
-                self.disband_at(escape_route)
+    def disband(self) -> None:
+        if self.position.is_friendly(self.player):
+            self.disband_at(self.position)
+        elif (escape_route := self.find_escape_route()) is not None:
+            self.disband_at(escape_route)
+        else:
+            logging.info(
+                f"No escape route available. Units were surrounded and destroyed "
+                "during transfer."
+            )
+            self.kill_all()
+
+    def should_disband(self, network: TransitNetwork) -> bool:
```

Comment:
This has the same naming confusion I mentioned before. This function shouldn't have any side effects. The _caller_ should be the one calling `disband`, not this function.

In other words, `validate_transfers` should be:

```
def disband_uncompletable_transfers(self) -> None:
    remaining = []
    for transfer in self.pending_transfers:
        if transfer.should_disband(self.network_for(transfer.position)):
            transfer.disband()
        else:
            remaining.append(transfer)
    self.pending_transfers = remaining
```

It makes `validate_transfers` marginally more complicated (no longer a one liner), but I can see what everything is doing based on the names without having to go digging into the implementation.

Aside from the misleading name this looks good though 👍

Reply:
i am so in love with these generators but i agree for the sake of readability i changed it as you said :) it's now even more clearer with `is_completable`